lib/generator.js

```diff
@@ -353,6 +353,9 @@ class Generator {
       if (!document) {
         const err = new Error('Input is not a correct AsyncAPI document so it cannot be processed.');
         err.diagnostics = diagnostics;
+        for (const diag of diagnostics) {
+          console.error('Diagnotic err:', diag);
```

Comment:
here you have a pretty good overview of diagnostics structure -> https://github.com/asyncapi/cli/blob/master/src/parser.ts#L73

I was wondering, how do you use generator? over CLI?

Reply:
I'm generally importing the generator and evoking it via a script, similar to the example I provide in https://github.com/asyncapi/generator/issues/1002#issuecomment-1724923172 (though I've updated to use `generate`, rather than `generateFromString`, since that was recently deprecated).

Well, if I'm tracing it properly, that Diagnostics structure is a fun wee jaunt from your link to AsyncAPI/cli, to [AsyncAPI Parser](https://github.com/asyncapi/parser-js/blob/master/src/types.ts), to a small extension in [Stoplight's Spectral/core](https://github.com/stoplightio/spectral/blob/1cae2245f0e89b980a8dcabb3eeeb9ae294508cd/packages/core/src/types/spectral.ts#L13), and finally ending up in [Stoplight's dianostic type](https://github.com/stoplightio/types/blob/master/src/diagnostics.ts) (with an interesting circular dependency to their [parser type](https://github.com/stoplightio/types/blob/master/src/parsers.ts) file for some of the Diagnostic Properties, as well). But I guess if we can rely on that being fairly stable, then we can use the more human-friendly summary.

I also tried tracing back exactly where the `diagnostics` object is created, to see if there was any variation in there I should be concerned about, but got a bit lost trying to figure out exactly what `NewParser` would be creating.

Anyway, I'll switch out just dumping the `diagnostics` object for the more human-friendly logging.

Reply:
@derberg Apologies for the delay, stuff came up in paid work. Change made.

It's interesting testing it against a `v2.6.0` AsyncAPI.yml definition, vs. a `v3.0.0` definition with a simple indentation error. `v2.6.0` was noisy to the point of me questioning if this was even going to be an effective change, whereas `v3.0.0` was comparatively terse and useful. (Payloads and repsonses listed below.) Obviously, in generally I assume it's recommended to try and use as current a version as possible, but backwards compatibility is always nice.

Listing line number can be slightly misleading, such as in my use case I listed above, where the initial format is first converted (in my case for the purpose of resolving external `$ref`s) and so line numbers can be lost/changed before parsing - but I'm willing to chalk that up to user error (especially in this case).

Overall, I think it's still better than the current lack of reporting, but not necessarily by much. Unsure how/if to proceed with this. Any thoughts?

Thanks, as always.

-----

3.0.0 payload.
```
asyncapi: '3.0.0'
info:
  title: Account Service 2
  version: 1.0.0
  description: This service is in charge of processing user signups
channels:
  user/signedup:
    subscribe:
      message:
        $ref: '#/components/messages/UserSignedUp'
components:
  messages:
    UserSignedUp:
        payload: #incorrect indentation error
        type: object
        properties:
          displayName:
            type: string
            description: Name of the user
          email:
            type: string
            format: email
            description: Email of the user

```
3.0.0 error logs:
```
Diagnostic err: Property "subscribe" is not expected to be here in path ["channels","user/signedup","subscribe"] starting L10 C18, ending L12 C54
Diagnostic err: "payload" property type must be object,boolean in path ["components","messages","UserSignedUp","payload"] starting L20 C19, ending L20 C23
Diagnostic err: Property "type" is not expected to be here in path ["components","messages","UserSignedUp","type"] starting L21 C16, ending L21 C24
Diagnostic err: Property "properties" is not expected to be here in path ["components","messages","UserSignedUp","properties"] starting L22 C21, ending L30 C46

/home/mball/dev/api-portal-alpha/node_modules/@asyncapi/generator/lib/generator.js:225
        const err = new Error('Input is not a correct AsyncAPI document so it cannot be processed.');
                    ^
Error: Input is not a correct AsyncAPI document so it cannot be processed.
```

2.6.0 payload:
```
asyncapi: '2.6.0'
info:
  title: Account Service 2
  version: 1.0.0
  description: This service is in charge of processing user signups
channels:
  user/signedup:
    subscribe:
      message:
        $ref: '#/components/messages/UserSignedUp'
components:
  messages:
    UserSignedUp:
        payload: #incorrect indentation error
        type: object
        properties:
          displayName:
            type: string
            description: Name of the user
          email:
            type: string
            format: email
            description: Email of the user

```
2.6.0 error logs:
```
Diagnostic err: AsyncAPI document should have "defaultContentType" field. in path [] starting L1 C0, ending L30 C46
Diagnostic err: AsyncAPI document should have "id" field. in path [] starting L1 C0, ending L30 C46
Diagnostic err: AsyncAPI document should have non-empty "servers" object. in path [] starting L1 C0, ending L30 C46
Diagnostic err: AsyncAPI object should have non-empty "tags" array. in path [] starting L1 C0, ending L30 C46
Diagnostic err: The latest version of AsyncAPi is not used. It is recommended update to the "3.0.0" version. in path ["asyncapi"] starting L2 C14, ending L2 C21
Diagnostic err: Info object should have "contact" object. in path ["info"] starting L3 C9, ending L6 C73
Diagnostic err: Info object should have "license" object. in path ["info"] starting L3 C9, ending L6 C73
Diagnostic err: Operation should have an "operationId" field defined. in path ["channels","user/signedup","subscribe"] starting L10 C18, ending L12 C54
Diagnostic err: "message" property must have required property "oneOf" in path ["channels","user/signedup","subscribe","message"] starting L11 C18, ending L12 C54
Diagnostic err: "message" property must match "then" schema in path ["channels","user/signedup","subscribe","message"] starting L11 C18, ending L12 C54
Diagnostic err: Property "payload" is not expected to be here in path ["channels","user/signedup","subscribe","message","payload"] starting L11 C18, ending L12 C54
Diagnostic err: "payload" property type must be object,boolean in path ["channels","user/signedup","subscribe","message","payload"] starting L11 C18, ending L12 C54
Diagnostic err: Property "properties" is not expected to be here in path ["channels","user/signedup","subscribe","message","properties"] starting L11 C18, ending L12 C54
Diagnostic err: Property "type" is not expected to be here in path ["channels","user/signedup","subscribe","message","type"] starting L11 C18, ending L12 C54
Diagnostic err: Message should have a "messageId" field defined. in path ["channels","user/signedup","subscribe","message"] starting L11 C18, ending L12 C54
Diagnostic err: "UserSignedUp" property must have required property "oneOf" in path ["components","messages","UserSignedUp"] starting L19 C21, ending L30 C46
Diagnostic err: "UserSignedUp" property must match "then" schema in path ["components","messages","UserSignedUp"] starting L19 C21, ending L30 C46
Diagnostic err: Message should have a "messageId" field defined. in path ["components","messages","UserSignedUp"] starting L19 C21, ending L30 C46
Diagnostic err: Property "payload" is not expected to be here in path ["components","messages","UserSignedUp","payload"] starting L20 C19, ending L20 C23
Diagnostic err: "payload" property type must be object,boolean in path ["components","messages","UserSignedUp","payload"] starting L20 C19, ending L20 C23
Diagnostic err: Property "type" is not expected to be here in path ["components","messages","UserSignedUp","type"] starting L21 C16, ending L21 C24
Diagnostic err: Property "properties" is not expected to be here in path ["components","messages","UserSignedUp","properties"] starting L22 C21, ending L30 C46

/home/mball/dev/api-portal-alpha/node_modules/@asyncapi/generator/lib/generator.js:225
        const err = new Error('Input is not a correct AsyncAPI document so it cannot be processed.');
                    ^
Error: Input is not a correct AsyncAPI document so it cannot be processed.

```

Reply:
lgtm